internal/core/api.rs

```diff
@@ -573,6 +573,36 @@ impl Window {
     }
 }
 
+#[cfg(feature = "rwh_06")]
+impl rwh_06::HasWindowHandle for Window {
+    fn window_handle<'a>(&'a self) -> Result<rwh_06::WindowHandle<'a>, rwh_06::HandleError> {
+        let adapter = self.0.window_adapter();
+        let wh = adapter.window_handle_06()?;
+        // Safety: The Rc<dyn WindowAdapter> owns this slint::Window (&self). Therefore the caller of
+        // this function must also have a strong reference to the window adapter. The adapter above
+        // was created from a self weak and is the same as the Rc<dyn WindowAdapter> of the caller.
```

Comment:
> The Rc<dyn WindowAdapter> owns this slint::Window

Is it for sure?
What if the Window is in a Cell or through another level of indirection? 
We know that this is the case for our backend, but we don't know what other backends are doing.

In particular: I can come up with this thing:

```rust
let some_other_window_adapter = ...
let window = slint::Window::new(Rc::downgrade(&some_other_window_adapter));
let rwa = window.window_handle();
some_other_window_adapter.hide();
drop(some_other_window_adapter);
// oops, pointers in rwa are dangling
do_something_with(rwa); // Undefined behavior
```

Or is that somehow prevented.

Alternative is to have a funtcion that return HasRawWindowHandle, but that API is not as elegant
```rust
fn window_handle_06(&self) -> impl HasRawWindowHandle {
   struct Wrapper(Rc<dyn WindowAdapter>);
   impl rwh_06::HasWindowHandle for Wrapper {
       fn window_handle<'a>(&'a self) -> Result<rwh_06::WindowHandle<'a>, rwh_06::HandleError> { 
           self.0.window_handle_06()
       }
   } 
   Wrapper(self.0.window_adapter())
}
```

Another alternative is to expose a function like `Window::adapter()` that return the Rc<dyn WindowAdapter>.  and then the user can call window_handle_06() on it.

Reply:
> > The Rc owns this slint::Window
> 
> Is it for sure? What if the Window is in a Cell or through another level of indirection? We know that this is the case for our backend, but we don't know what other backends are doing.

I think as long as we have this function in our `WindowAdapter` trait, it can't be an indirection that wouldn't make this safe:

```rust
pub trait WindowAdapter {
    /// Returns the window API.
    fn window(&self) -> &Window;
```

> In particular: I can come up with this thing:
> 
> ```rust
> let some_other_window_adapter = ...
> let window = slint::Window::new(Rc::downgrade(&some_other_window_adapter));
> let rwa = window.window_handle();
> some_other_window_adapter.hide();
> drop(some_other_window_adapter);
> // oops, pointers in rwa are dangling
> do_something_with(rwa); // Undefined behavior
> ```
> 
> Or is that somehow prevented.

This panics, as we call `self.window_adapter_weak.upgrade().unwrap()`.
 
> Alternative is to have a funtcion that return HasRawWindowHandle, but that API is not as elegant
> 
> ```rust
> fn window_handle_06(&self) -> impl HasRawWindowHandle {
>    struct Wrapper(Rc<dyn WindowAdapter>);
>    impl rwh_06::HasWindowHandle for Wrapper {
>        fn window_handle<'a>(&'a self) -> Result<rwh_06::WindowHandle<'a>, rwh_06::HandleError> { 
>            self.0.window_handle_06()
>        }
>    } 
>    Wrapper(self.0.window_adapter())
> }
> ```
> 
> Another alternative is to expose a function like `Window::adapter()` that return the Rc. and then the user can call window_handle_06() on it.

Yes, those are two alternatives. I think the moment we return a wrapper, we basically allow people to hold a reference to the plain `Rc<dyn WindowAdapter>`.  We could do that, but I'd rather avoid it if we can.

Reply:
> This panics, as we call self.window_adapter_weak.upgrade().unwrap().

Where does this panics.  The other window adapter is only going to be drop'ed *after* the call to window_handle()

Reply:
You're right, I misread the example. Ouch.